lib/converter.js

```diff
@@ -1,133 +1,94 @@
-/* eslint-env phantomjs */
-/* eslint-disable no-console, no-var, prefer-arrow-callback, object-shorthand */
-"use strict";
-
-var webpage = require("webpage");
-var system = require("system");
-
-var HTML_PREFIX = "<!DOCTYPE html><style>html, body { margin: 0; padding: 0; } " +
-                  "svg { position: absolute; top: 0; left: 0; }</style>";
-
-if (system.args.length !== 2) {
-    console.error("Usage: converter.js options");
-    phantom.exit();
-} else {
-    convert(system.args[1]);
-}
-
-function convert(options) {
-    try {
-        options = JSON.parse(options);
-    } catch (e) {
-        console.error("Unable to parse options.");
-        console.error(e);
-        phantom.exit();
-        return;
-    }
-
-    var page = webpage.create();
-
-    var source = "";
-    while (!system.stdin.atEnd()) {
-        source += system.stdin.readLine() + "\n";
-    }
-
-    page.onLoadFinished = function (status) {
-        if (status !== "success") {
-            console.error("Unable to load the source file.");
-            phantom.exit();
-            return;
-        }
+/* eslint-env browser */
 
-        try {
-            if (options.width !== undefined || options.height !== undefined) {
-                setSVGDimensions(page, options.width, options.height);
-            }
+"use strict";
 
-            var dimensions = getSVGDimensions(page);
+exports.convert = (page, options) => {
+    return setSVGDimensions(page, options)
+        .then(() => getSVGDimensions(page))
+        .then(dimensions => {
             if (!dimensions) {
-                console.error("Width or height could not be determined from either the source file or the supplied " +
-                              "dimensions");
-                phantom.exit();
-                return;
+                throw new Error("Width or height could not be determined from either the source file or the supplied " +
+                                "dimensions");
             }
 
-            setSVGDimensions(page, dimensions.width, dimensions.height);
-
-            page.viewportSize = {
-                width: dimensions.width,
-                height: dimensions.height
-            };
-            page.clipRect = {
-                top: 0,
-                left: 0,
-                width: dimensions.width,
-                height: dimensions.height
+            return setSVGDimensions(page, dimensions);
+        })
+        .then(() => createPNGDataURL(page));
+};
+
+function createPNGDataURL(page) {
```

Comment:
So the pipeline here is to draw the SVG to a canvas then get a data URL for a png and then base64 decode it into binary? That seems terrible :(.

Why can't we just use an `<img src="svg.svg">` (or navigate Chromium to the SVG directly) and resize the window appropriately, then use Pupetteer's screenshot API?

We should also be able to avoid a lot (not all) of the SVG dimensions stuff if we can just use the input width and height to set the window width/height. Remember that per https://github.com/domenic/svg2png#exact-resizing-behavior the intent is "to stick as close to the behavior of loading a SVG file in your browser as possible"

Reply:
The problem with your suggestion of having ` <img src="svg.svg">` is that, when the API is directly passed a SVG buffer, we'd have to write it to a temporary file so that it could be imported in such a way. I initial ignored the screenshot approach as I assumed it would break for SVGs containing transparency, but I may have been wrong after looking at their docs.

The temporary file writing could impact performance, but I'm not sure how important that is for this library. I doubt anyone is using it in application "hot paths". I assume it's mostly used in build processes or ad-hoc commands. I could write it in a way that the API reads directly from a file when a string is passed, otherwise treats it as a buffer and writes it to a temporary file.

I'll investigate the screenshot approach further and see how I get on. However, I'm not convinced that this will work around the issue of external files in cases where a buffer is passed directly to the API, but I'll check that as well.

Reply:
In another project (using chromy), I've had success passing the SVG as a data-uri to open/navigate  to; also screenshotting with `svg` selector works well.

Reply:
Ah, good point, I would prefer to avoid temporary files. I'm not sure what the right path is then; data URL, or write the bytes directly to canvas? (via Buffer -> Uint8Array -> ArrayBuffer -> Uint8ClampedArray, none of which involve a copy, just wrapping new types around the same binary data)